projects/opendr_ws_2/src/opendr_perception/opendr_perception/object_tracking_2d_deep_sort_node.py

```diff
@@ -0,0 +1,246 @@
+
+        self.get_logger().info("Using model_name: {}".format(model_name))
+
+        self.detector = detector
+        self.learner = ObjectTracking2DDeepSortLearner(
+            device=device, temp_path=temp_dir,
+        )
+        if not os.path.exists(os.path.join(temp_dir, model_name)):
+            ObjectTracking2DDeepSortLearner.download(model_name, temp_dir)
+
+        self.learner.load(os.path.join(temp_dir, model_name), verbose=True)
+
+        # Initialize OpenDR ROSBridge object
+        self.bridge = ROS2Bridge()
+        self.tracking_id_publisher = self.create_publisher(
+            Int32MultiArray, output_tracking_id_topic, 1
+        )
+
+        if output_image_topic is not None:
+            self.output_image_publisher = self.create_publisher(
+                ROS_Image, output_image_topic, 1
+            )
+
+        self.detection_publisher = self.create_publisher(
+            Detection2DArray, output_detection_topic, 1
+        )
+
+        self.create_subscription(ROS_Image, input_image_topic, self.callback, 1)
+
+    def callback(self, data):
+        """
+        Callback that process the input data and publishes to the corresponding topics
+        :param data: input message
+        :type data: sensor_msgs.msg.Image
+        """
+
+        # Convert sensor_msgs.msg.Image into OpenDR Image
+        image = self.bridge.from_ros_image(data, encoding="bgr8")
+        detection_boxes = self.detector.infer(image)
+        image_with_detections = ImageWithDetections(image.numpy(), detection_boxes)
+        tracking_boxes = self.learner.infer(image_with_detections)
+
+        if self.output_image_publisher is not None:
+            frame = image.opencv()
+            draw_predictions(frame, tracking_boxes)
+            message = self.bridge.to_ros_image(
+                Image(frame), encoding="bgr8"
+            )
+            self.output_image_publisher.publish(message)
+            self.get_logger().info("Published annotated image")
+
+        ids = [int(tracking_box.id) for tracking_box in tracking_boxes]
+
+        # Convert detected boxes to ROS type and publish
+        ros_boxes = self.bridge.to_ros_boxes(detection_boxes)
+        if self.detection_publisher is not None:
+            self.detection_publisher.publish(ros_boxes)
+            self.get_logger().info("Published " + str(len(detection_boxes)) + " detection boxes")
+
+        ros_ids = Int32MultiArray()
+        ros_ids.data = ids
+
+        if self.tracking_id_publisher is not None:
+            self.tracking_id_publisher.publish(ros_ids)
+            self.get_logger().info("Published " + str(len(ids)) + " tracking ids")
+
+
+colors = [
+    (255, 0, 255),
+    (0, 0, 255),
+    (0, 255, 0),
+    (255, 0, 0),
+    (35, 69, 55),
+    (43, 63, 54),
+]
+
+
+def draw_predictions(frame, predictions: TrackingAnnotation, is_centered=False, is_flipped_xy=True):
+    global colors
+    w, h, _ = frame.shape
+
+    for prediction in predictions.boxes:
```

Comment:
I quote the comment from ROS1 node here:

> On line 147 i get a warning for predictions.boxes as the IDE can't find a reference to .boxes in target.TrackingAnnotation that is type hinted in the arguments of the method. There is a .boxes in TrackingAnnotationList, though. I cannot test if this actually problematic on runtime, but the method seems to be provided with a python list from the return of learner.infer in line 106, which includes predictions from the tracker, which themselves are TrackingAnnotationList or an element of a TrackingAnnotationList if is_single_image is true within infer. So all in all, predictions can be a list with TrackingAnnotationList elements, so it should not have a .boxes itself, but a .boxes for each of its elements, if infer is provided with single images, otherwise this would produce an error, but i can't say if it is possible in the ROS node. Nevertheless, i feel that this should be cleared up, through the type hint or otherwise.

Reply:
Same as in ROS1